src/Core/Services/BuildRequestStateMiddleware.cs

```diff
@@ -26,14 +35,92 @@ public BuildRequestStateMiddleware(RequestDelegate next)
+
+        using Activity? activity = !isIntrospectionQuery ?
+            TelemetryTracesHelper.DABActivitySource.StartActivity($"{method} /{route}") : null;
+
+        try
         {
-            // Because Request.Headers is a NameValueCollection type, key not found will return StringValues.Empty and not an exception.
-            StringValues clientRoleHeader = httpContext.Request.Headers[AuthorizationResolver.CLIENT_ROLE_HEADER];
-            context.ContextData.TryAdd(key: AuthorizationResolver.CLIENT_ROLE_HEADER, value: clientRoleHeader);
+            // We want to ignore introspection queries DAB uses to check access to GraphQL since they are not sent by the user.
+            if (!isIntrospectionQuery)
+            {
+                TelemetryMetricsHelper.IncrementActiveRequests(apiType);
+                if (activity is not null)
+                {
+                    activity.TrackRestControllerActivityStarted(
+                        httpMethod: method,
+                        userAgent: httpContext.Request.Headers["User-Agent"].ToString(),
+                        actionType: (context.Request.Query!.ToString().Contains("mutation") ? OperationType.Mutation : OperationType.Query).ToString(),
+                        httpURL: string.Empty, // GraphQL has no route
+                        queryString: null, // GraphQL has no query-string
+                        userRole: httpContext.Request.Headers["X-MS-API-ROLE"].FirstOrDefault() ?? httpContext.User.FindFirst("role")?.Value,
+                        apiType: apiType);
+                }
+            }
+
+            await InvokeAsync();
+        }
+        finally
+        {
+            stopwatch.Stop();
+
+            HttpStatusCode statusCode;
+
+            // We want to ignore introspection queries DAB uses to check access to GraphQL since they are not sent by the user.
+            if (!isIntrospectionQuery)
+            {
+                // There is an error in GraphQL when ContextData is not null
+                if (context.Result!.ContextData is not null)
+                {
+                    if (context.Result.ContextData.ContainsKey(WellKnownContextData.ValidationErrors))
+                    {
+                        statusCode = HttpStatusCode.BadRequest;
+                    }
+                    else if (context.Result.ContextData.ContainsKey(WellKnownContextData.OperationNotAllowed))
+                    {
+                        statusCode = HttpStatusCode.MethodNotAllowed;
+                    }
+                    else
+                    {
+                        statusCode = HttpStatusCode.InternalServerError;
+                    }
+
+                    string errorMessage = context.Result.Errors![0].Message;
+                    Exception ex = new(errorMessage);
+
+                    // Activity will track error
+                    activity?.TrackRestControllerActivityFinishedWithException(ex, statusCode);
```

Comment:
nit: the name `TrackRestControllerActivityFinishedWithException` is misleading since this is now being used for GraphQL too.. Consider using a generic TrackController.. name